--- java/engine/org/apache/derby/iapi/jdbc/BrokeredPreparedStatement.java.sav	2004-11-30 20:20:48.000000000 +0100
+++ java/engine/org/apache/derby/iapi/jdbc/BrokeredPreparedStatement.java	2005-02-09 15:54:23.974922304 +0100
@@ -502,6 +502,19 @@
 		return getPreparedStatement();
 	}
 
+        public void setURL(int i, java.net.URL x)
+        throws SQLException
+    {
+        getPreparedStatement().setURL( i, x);
+    }
+
+    public ParameterMetaData getParameterMetaData()
+        throws SQLException
+    {
+        return getPreparedStatement().getParameterMetaData();
+    }
+
+
 	/**
 		Create a duplicate PreparedStatement to this, including state, from the passed in Connection.
 	*/
