--- java/engine/org/apache/derby/impl/jdbc/EmbedCallableStatement20.java.sav	2004-11-30 20:24:14.000000000 +0100
+++ java/engine/org/apache/derby/impl/jdbc/EmbedCallableStatement20.java	2005-02-09 16:36:01.972169000 +0100
@@ -22,6 +22,7 @@
 
 import java.math.BigDecimal;
 import java.sql.CallableStatement;
+import java.sql.ParameterMetaData;
 import java.sql.SQLException;
 import java.sql.Date;
 import java.sql.Time;
@@ -1159,6 +1160,18 @@
 	{
 		throw Util.notImplemented();
 	}
+
+        public ParameterMetaData getParameterMetaData()
+    throws SQLException
+        {
+                checkStatus();
+                if (preparedStatement == null)
+                        return null;
+
+                return new EmbedParameterMetaData30(
+                                getParms(), preparedStatement.getParameterTypes());
+        }
+
 }
 
 
